// @automock/jest/dist/lib/test-bed-resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestBedResolver = void 0;
require("reflect-metadata");
const mock_resolver_1 = require("./mock-resolver");
class TestBedResolver {
    constructor(reflector, mockFn, targetClass) {
        this.reflector = reflector;
        this.mockFn = mockFn;
        this.targetClass = targetClass;
        this.dependencies = new Map();
        this.depNamesToMocks = new Map();
        this.dependencies = this.reflector.reflectDependencies(targetClass);
    }
    mock(typeOrToken) {
        return {
            using: (mockImplementation) => {
                this.depNamesToMocks.set(typeOrToken, this.mockFn(mockImplementation));
                return this;
            },
        };
    }
    compile() {
        const deps = this.mockUnMockedDependencies();
        const values = Array.from(deps.values());
        return {
            unit: new this.targetClass(...values),
            unitRef: new mock_resolver_1.MockResolver(deps),
        };
    }
    mockUnMockedDependencies() {
        const map = new Map();
        for (const [key, dependency] of this.dependencies.entries()) {
            const overriddenDep = this.depNamesToMocks.get(key);
            const mock = overriddenDep ? overriddenDep : this.mockFn();
            map.set(key, mock);
        }
        return map;
    }
}
exports.TestBedResolver = TestBedResolver;
